tachyon/math/elliptic_curves/msm/algorithms/icicle/icicle_msm.h

```diff
@@ -167,14 +167,32 @@ bool IcicleMSM<bn254::G1AffinePoint>::Run(const BaseContainer& cpu_bases,
     return false;
   }
 
-  ::bn254::projective_t ret;
-  gpuError_t error = tachyon_bn254_g1_msm_cuda(
-      reinterpret_cast<const ::bn254::scalar_t*>(std::data(cpu_scalars)),
-      reinterpret_cast<const ::bn254::affine_t*>(std::data(cpu_bases)),
-      bases_size, *config_, &ret);
-  if (error != gpuSuccess) return false;
-  ret = ::bn254::projective_t::to_montgomery(ret);
-  *cpu_result = base::bit_cast<ProjectivePoint<Curve>>(ret);
+  size_t bitsize = static_cast<size_t>(
+      (config_->bitsize == 0) ? ::bn254::scalar_t::NBITS : config_->bitsize);
+
+  size_t divisions = DetermineMsmDivisionsForMemory(
+      sizeof(::bn254::scalar_t), sizeof(::bn254::affine_t),
+      sizeof(::bn254::projective_t), bases_size, config_->c, bitsize,
+      static_cast<size_t>(config_->precompute_factor),
+      static_cast<size_t>(config_->batch_size));
```

Comment:
Please call this function only once during initialization! It's not a cheap operation, and in most cases, the available memory will remain similar throughout the program's execution.

Reply:
In that case, free memory is also checked only when initializing. Won't free memory fluctuate significantly during program execution? If there is a process that occupies a large amount of free memory, memory limitation issues may occur.

Reply:
In production, I believe this situation rarely happens. The machine that produces a ZK proof probably will be ran with this purpose only!